### PR TITLE
MRG: Add label center_of_mass

### LINKED ISSUE #3293: DOC: Clarify source space functions
We should make it clear that `mne.setup_source_space` sets up a standard-pipeline surface-based `SourceSpaces` object, and setup cross-references to `mne.setup_volume_source_space`

### PR DESCRIPTION
**Pull request by Eric89GXL**
@nfoti can you cross-check that it gives the same COM vertex numbers as your method?

Takes care of part of #3307.

Ready for review/merge from my end.

Closes #3293.

### CODE REVIEW

**Comment by nfoti** [comment]
I am not getting the same center of mass as my method. For subject `'sample'` and label `'bankssts-lh'` my code returns vertex `60530` as the center while the new `center_of_mass` function returns vertex `60526`. These vertices are close to each other, but not the same.

Just to clarify that I'm using this function correctly, I do the following:
```
        labels[i].values = np.ones_like(labels[i].values)
        com = labels[i].center_of_mass(subject='sample',
                                       subjects_dir=subject_dir,
                                       restrict_vertices=True)
```
where `labels[i]` is a list containing the `Label`s for the subject. To be thorough I'm copying my function here which is derived from the `_select_source_in_label` function in `mne/simulation/source.py`.
```
def select_center_vertex_in_label(src, label):
    """ Select vertex closest (in Euclidean distance) to the center of mass of
        the given label.

    Parameters
    ----------
    src : list of dict
        The source space
    label : Label
        the label (read with mne.read_label)

    Returns
    -------
    center_vert : int
        index of central vertex
    """

    pos = label.pos
    mean_pos = np.mean(pos, axis=0)
    dists = np.sqrt(np.sum((pos - mean_pos)**2, axis=1))

    hemi_to_ind = {'lh': 0, 'rh': 1}
    src_sel = np.intersect1d(src[hemi_to_ind[label.hemi]]['vertno'], label.vertices)
    idx_sel = np.searchsorted(label.vertices, src_sel)
    center_vertex = idx_sel[np.argmin(dists[idx_sel])]
    return center_vertex
```

Also, it looks like this function returns the index of the center of mass in the source spaces vertex list, however, since we're computing the center of mass for a label it seems more likely that a user would want the index of the center of mass in the `Label`'s vertex list. Perhaps if `restrict_vertices` is `True` then we return the index from the `Label`'s vertex list (i.e. the index so that `label.vertices[com]` is the center of mass) and if `restrict_vertices` is `False` then we return the index from the source space vertex list.

**Comment by nfoti** [comment]
It looks like the big difference is that @Eric89GXL is computing the center of mass based on the activity of the vertices, whereas I'm computing it with the position of the vertices.

**Comment by Eric89GXL** [comment]
Yeah your call is fine.

> Also, it looks like this function returns the index of the center of mass in the source spaces vertex list, however, since we're computing the center of mass for a label it seems more likely that a user would want the index of the center of mass in the Label's vertex list.

It's returning the index in the high-density mesh, which is where we tend to give indices to when possible. This behavior is also maximally consistent with the existing `stc.center_of_mass` method so I don't think we can change it. I can add a note in the docstring that the corresponding index in the label would be given by `np.where(label.vertices == idx)[0][0]` to help.

I see two potential differences between yours and mine:

1. Mine uses `label.values` to weight, yours uses uniform weighting.
2. Mine uses positions from the "sphere" surface, whereas yours uses them from `label.pos`. I'm actually not sure offhand where `label.pos` comes from. I thought it would be "sphere" but maybe it's "white". I actually think "sphere" might make the most sense here, but I'm not sure.

Can you modify the branch you checked out locally to load the "white" surface instead of "sphere" and see if it gives the same results?

File: mne/source_estimate.py [hunk 1]
```diff
@@ -910,6 +909,24 @@ def transform(self, func, idx=None, tmin=None, tmax=None, copy=False):
...
+def _center_of_mass(vertices, values, hemi, subject, subjects_dir,
+                    restrict_vertices):
+    """Helper to find the center of mass on a surface"""
+    subjects_dir = get_subjects_dir(subjects_dir, raise_error=True)
+    surf = read_surface(op.join(subjects_dir, subject, 'surf',
+                                hemi + '.sphere'))
+    if restrict_vertices is True:
+        restrict_vertices = vertices
+    elif restrict_vertices is False:
+        restrict_vertices = np.arange(surf[0].shape[0])
```
Comment (agramfort) [inline mne/source_estimate.py]: if restrict_vertices is not a bool BOOM

File: mne/source_estimate.py [hunk 2]
```diff
@@ -1195,14 +1212,18 @@ def extract_label_time_course(self, labels, src, mode='mean_flip',
...
                        subjects_dir=None):
-        """Return the vertex on a given surface that is at the center of mass
-        of  the activity in stc. Note that all activity must occur in a single
-        hemisphere, otherwise an error is returned. The "mass" of each point in
-        space for computing the spatial center of mass is computed by summing
-        across time, and vice-versa for each point in time in computing the
-        temporal center of mass. This is useful for quantifying spatio-temporal
-        cluster locations, especially when combined with the function
-        mne.source_space.vertex_to_mni().
+        """Compute the the center of mass of activity
```
Comment (agramfort) [inline mne/source_estimate.py]: the the

**Comment by agramfort** [comment]
that's it for me

File: mne/source_estimate.py (hunk 1 above)
Comment (Eric89GXL) [inline mne/source_estimate.py]: Actually it is designed that way, it can be a `ndarray`. I guess we could do some input checking on the ndarray if you want?

Comment (Eric89GXL) [inline mne/source_estimate.py]: ...but maybe we should get rid of `ndarray` because 1) originally I had the same reaction as you, 2) I *highly* doubt anyone is actually using it, and 3) the same thing can be achieved by modifying the label itself properly (although it is a bit more annoying)

**Comment by Eric89GXL** [comment]
@agramfort comments addressed

@nfoti can you try it with `surf='white'` to see if the results now match yours?

**Comment by agramfort** [comment]
merge if @nfoti is happy

**Comment by nfoti** [comment]
The results still don't match using `surf='white'`. For the `bankssts-lh` label my method yields vertex `60530`, while your method with `surf='white'` yield vertex `59441` and with `surf='sphere'` yields vertex `60525`.

It does seem like `label.pos` isn't the same as either `sphere` or `white` though. Given this, it's probably best not use it and to instead load one of the surfaces like you do.

**Comment by Eric89GXL** [comment]
Hmm, based on the `read_labels_from_annot`, if you are using that, they should be on the `white` surface. Do the obtained vertices look reasonable at least?

**Comment by nfoti** [comment]
@Eric89GXL : Yes, they look reasonable. Are we sure we're using the same structural data? The subjects haven't been processed again since you sent them to me, have they?

**Comment by Eric89GXL** [comment]
The structurals don't really change over time, or at least shouldn't have. There was a bug in the forward that probably made some minor changes to the forward on my machine, but really neither of those matter. They won't affect the consistency *within your system* of where the surface values say that they are, i.e. you shouldn't get different vertex numbers if the algorithms are equivalent. I'll have to look and see why they are not.

**Comment by dgwakeman** [comment]
They will change though, if they are re-processed with a different version of FreeSurfer (just fyi)

**Comment by Eric89GXL** [comment]
In this case we only ever process them once fortunately

**Comment by Eric89GXL** [comment]
@nfoti I think I found the difference using [this gist] (https://gist.github.com/Eric89GXL/498ae0684f6fec8ee2d460b6120e74b5) -- your routine seems to restrict results to being within the label and within the source space, whereas mine doesn't by default. Forcing it to find a point both in the source space and in the label (with appropriate `restrict_vertices`) makes it work. Also, I suspect your routine does ~50x as many distances as it needs to assuming a high-density label, since it computes the distances for all label vertices, but can only ever sub-select the vertices that also are in the source space.

TL;DR: If you use `restrict_vertices = np.intersect1d(src[idx]['vertno'], label.vertices)` you should find equivalence between our methods

**Comment by nfoti** [comment]
@Eric89GXL Using your definition of `restrict_vertices` I get the same center as you.

Since a motivating use-case for this function is to find the center vertex in a label with respect to a source space can we either (i) allow a source space to be passed in and if so set `restrict_vertices` to force the result to be in the source space, or (ii) add the above definition of `restrict_vertices` to do this to the documentation in a conspicuous place?

**Comment by Eric89GXL** [comment]
We could change it so that `restrict_vertices` itself can take a `SourceSpaces` instance. That could be nice. I'll try coding it.

**Comment by nfoti** [comment]
Awesome. Looking forward to it.

**Comment by nfoti** [comment]
Seems to work well. The only thing is that I'm getting different centers of mass from my method and your method for the `cuneus-lh` label. What's strange is that when I use the `restrict_vertices = np.intersect1d(...)` as you suggest above both your and my method yield the same results, but when I pass in a source space I get different values, but just for that label.

Otherwise, I think this is really easy to use and is great.

**Comment by Eric89GXL** [comment]
Can you debug to see if two points have very similar distances from the true COM for that case? That could cause it to be different from one method to the other due to numerical error differences.

**Comment by Eric89GXL** [comment]
Since @nfoti is happy I'll go ahead and merge, assuming the difference in the one label arises due to numerical differences. @nfoti let me know if you find otherwise and I'll fix it.

**Comment by nfoti** [comment]
I just checked the positions of the points returned and the difference is not just numerical. The position of the point returned by `center_of_mass` is `[-10.89, -71.5, 21.71]` and the position of the point returned by my function is `[-8.1, -74.15, 23.93]`. So they're quite different. This is really strange as it only happens for the one label (`cuneus-lh`) when passing in the source space rather than specifying `restrict_vertices`.

**Comment by Eric89GXL** [comment]
> The position of the point returned by center_of_mass is [-10.89, -71.5, 21.71] and the position of the point returned by my function is [-8.1, -74.15, 23.93]

I don't mean look at the position of the returned points relative to one another (which I think is what you posted here) but rather:

>> ... if two points have very similar distances from the true COM ...

In other words, I'm thinking you could check what the actual ideal center-of-mass locations are for the two functions (they might be similar but not identical), and then also look at the `np.sort` of the distances from the eligible vertices to those. If my algorithm thinks the best two distances are from vertices *i* and *j* are 2.4999999 and 2.5000001 mm and yours thinks that the order is actually *j* then *i* due to numerical errors (order of operations, etc. differing between them), then it could result in points that are that far away from one another. Those two points are only 4.4 mm from one another, so could be neighbors in the decimated space (I think?).

But thinking about it more, doing `restrict_vertices=src` is not meant to be equivalent to `restrict_vertices=np.intersect1d(src[0]['vertno'], label.vertices)`. The former has one restriction, namely that the vertex has to land in the decimated source space. The latter has two restrictions, the same as the former *plus* it must also be in the label. Is the point that the `restrict_vertices=src` finds outside the label? I think this is reasonable behavior (thinking about e.g. where the COM for an annulus-shaped label should land in both conditions) and would explain the difference.
